casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala

```diff
@@ -147,50 +135,85 @@ class RuntimeManager private (val emptyStateHash: ByteString, runtimeContainer:
     } yield (result.patterns, result.continuation.taggedCont.parBody.get.body)
   }
 
-  private def getResetRuntime(hash: StateHash) = {
-    val runtime   = runtimeContainer.take()
-    val blakeHash = Blake2b256Hash.fromByteArray(hash.toByteArray)
-    Try(runtime.space.reset(blakeHash)) match {
-      case Success(_) => runtime
-      case Failure(ex) =>
-        runtimeContainer.put(runtime)
-        throw ex
-    }
+  private def newEval(terms: Seq[Deploy], runtime: Runtime, initHash: StateHash)(
+      implicit scheduler: Scheduler): (StateHash, Seq[InternalProcessedDeploy]) = {
+    implicit val costAccountingAlg = CostAccountingAlg.unsafe[Task](CostAccount.zero)
+
+    def doEval(terms: Seq[Deploy],
+               hash: Blake2b256Hash,
+               acc: Vector[InternalProcessedDeploy]): (StateHash, Vector[InternalProcessedDeploy]) =
+      terms match {
+        case deploy +: rem =>
+          runtime.space.reset(hash)
+          val (cost, errors) = injAttempt(deploy, runtime.reducer, runtime.errorLog)
+          val newCheckpoint  = runtime.space.createCheckpoint()
+          val deployResult = InternalProcessedDeploy(deploy,
+                                                     cost,
+                                                     newCheckpoint.log,
+                                                     DeployStatus.fromErrors(errors))
+
+          if (errors.isEmpty) doEval(rem, newCheckpoint.root, acc :+ deployResult)
+          else doEval(rem, hash, acc :+ deployResult)
+
+        case _ => (ByteString.copyFrom(hash.bytes.toArray), acc)
+      }
+
+    doEval(terms, Blake2b256Hash.fromByteArray(initHash.toByteArray), Vector.empty)
+  }
+
+  private def replayEval(
+      terms: Seq[InternalProcessedDeploy],
+      runtime: Runtime,
+      initHash: StateHash)(implicit scheduler: Scheduler): Either[Throwable, StateHash] = {
+    implicit val costAccountingAlg = CostAccountingAlg.unsafe[Task](CostAccount.zero)
+
+    def doReplayEval(terms: Seq[InternalProcessedDeploy],
+                     hash: Blake2b256Hash): Either[Throwable, StateHash] =
+      terms match {
+        case InternalProcessedDeploy(deploy, _, log, status) +: rem =>
+          runtime.replaySpace.rig(hash, log.toList)
```

Comment:
Yes, that is a good observation, this does inherently serialize the deploys in the block. I was hesitant to do that at first, but here is why I decided to stick with it:
1. We don't really have parallel deploys now anyways because there is a single hot store and only one deploy can be using it at a time.
2. Some deploys do legitimately need to be serialized because they do not commute with one another (the final result differs depending on the order the deploys were executed in). So to do proper parallel deploys you would actually have groups of deploys which are unordered with respect to each other and a serial set within each group. That would be a more complicated change than I wanted to do here and I think it would work better once we have a native rspace merge operation (see https://rchain.atlassian.net/browse/CORE-1010). Also this lays some ground work for that because this would be the structure of a single serialized group and in a future PR it would only remain to introduce the possibility of multiple such independent groups.

Reply:
Consider the following example where it is necessary to communicate which deploy was executed first.

Suppose that the following term exists in the tuplespace:
```
for(@x <- @"x"){
  @"y"!(x + 1) |
  for(@x <- @"x"){ @"z"!("Hello, $name" %% {name: x}) }
}
```
And there are two deploys to be executed: D1 = `@"x"!(7)` and D2 = `@"x"!("Alice")`. If D1 goes first and D2 second then they both complete without error. However, if D2 goes first and then D1 second then D2 will fail and D1 will succeed. Note that in both cases D1 has a comm event with the outer for because state changes occurring in error deploys are rolled back.

If the first node runs D2 then D1 and a second node tries to replay the results but does D1 then D2, it will conclude that the block is invalid because it claimed an error where there is none.

Reply:
The node that replays D1 and D2 will not be able to replay D2 so it will be fine.

Reply:
As in when D2 goes first there will be an explicit error but when D1 goes first there will be an untraced comm event.

Reply:
And we don't care how ErroredDeploys (see below) actually error as long as they error.

Reply:
there is a single hot store and only one deploy can be using it at a time. - It is fine that there is a single hot sore. The idea is that you can have that single hot store be running multiple reductions at once. However it currently doesn't do that but that is rspace's implementation details. They should be able to swap it out under us into a parallel implementation and we should see a magical performance boost.

Reply:
Getting an untraced comm event is bad though. I consider that an internal error as opposed to a user error because otherwise we can't tell what is a faulty log and what is just a error. So replaying D1 first then getting an untraced comm event still makes the block invalid.

This example also shows why you can't just do all the errors at the end; D2 would no longer be an error if it came after D1 (except it would be an untraced comm event so there would an error, but not a user one).

And we do have parallelism in the evaluation of a deploy I think. As in I think the Rholang reducer is multithreaded. To get more of a boost we would need to identify what deploys could be safely run in parallel (i.e. commute with each other) on our end then kick off multiple reducers.

Reply:
I would say that paralleling deploys make sense. From the user perspective things that go into deploy should not have to know about each other. If the deployer deploys two contracts that need to be evaluated in specific order (as in the example @birchmd gave) it should be their job to do so and the only way would be to do them in separate blocks since rholang is inherently non-deterministic and we cannot guarantee any execution order
> `@"x"!(7) | @"x"!(Alice)`
Can be evaluated in any order right?

Reply:
The issue is that we can execute deploys in parallel fairly easily but we're having difficulty trying to come up with a way to replay the once executed deploys in parallel (for validating blocks) given how deploys with user errors still have to be kept around.

Reply:
Well, if errored deploys should not influence the tuplespace state doing it as you proposed should work -- do errored deploys in the end and validate that they didn't change the poststate.

Reply:
The issue is that errored deploys do influence the state. Also note that you don't know if a deploy contains an error until you run it once. There seems to be only two options: a) create a checkpoint before trying each deploy so that if you encounter an error on the next deploy you can rollback to that checkpoint or b) just run all deploys until you encounter and error and then just rollback to the previous block checkpoint and retry again with that erroneous deploy removed.

Reply:
> The issue is that errored deploys do influence the state.

How is that? Aren't we rolling back to the hash before the replaying the errored deploy? I am missing something.

Reply:
> Aren't we rolling back to the hash before the replaying the errored deploy?

Yeah that is what I mean when I said the below

a) create a checkpoint before trying each deploy so that if you encounter an error on the next deploy you can rollback to that checkpoint

Reply:
But note that a) means that you're creating a checkpoint after each deploy which is slow.

Reply:
So it doesn't change the tuplespace, it requires some action exactly in order to not change the tuplspace.